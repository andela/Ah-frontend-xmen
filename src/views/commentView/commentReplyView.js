import React, { Component } from 'react';
import CommentReplyBanner from '../../components/comments/commentReplyBanner';
import { getCommentReplies } from '../../actions/commentActions/commentReplyActions';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { createCommentReplyAction, deleteCommentReplyAction } from '../../actions/commentActions/commentReplyActions';
import {ToastContainer} from 'react-toastify';

class CommentReplyView extends Component {

  state = {
    reply_body:""
  }

  handleChange = e => {
   this.setState({
     reply_body:e.target.value
   })
  };
  handleSubmit = e => {
    e.preventDefault()
    this.props.createCommentReplyAction({
      slug: this.props.match.params.slug,
      commentId: this.props.commentId,
      reply_body: this.state
    })

    setTimeout(()=>{
      this.props.getCommentReplies({
        slug:this.props.match.params.slug,
        commentId:this.props.commentId
      })
    },500)

  }

  handleFetch = (e) => {
    this.props.getCommentReplies({
      slug: this.props.match.params.slug,
      commentId: this.props.commentId
    })
  }
  handleDelete = (e) => {
    let replies = this.props.commentReplyState.payload

    this.props.deleteCommentReplyAction({
      slug: this.props.match.params.slug,
      commentId: this.props.commentId,
      reply_id: e.target.id
    })

    // setTimeout(()=>{
    //   // this.props.getCommentReplies({
    //   //   slug: this.props.match.params.slug,
    //   //   commentId: this.props.commentId
    //   // })
    //   window.location.reload()
    // },2000)

  }

  componentWillReceiveProps(nextProps, nextContext) {
    console.log(nextProps.commentReplyState.replyDeleteSuccess)
    if(nextProps.commentReplyState.isSuccessful){
      this.setState({
        reply_body:""
      })
    }
    else if(nextProps.commentReplyState.replyDeleteSuccess){
      this.props.getCommentReplies({
        slug: this.props.match.params.slug,
        commentId: this.props.commentId
      })
    }


  }
  componentWillUpdate(nextProps, nextState, nextContext) {
    console.log('will update')
  }

  render() {
    const replies = this.props.commentReplyState.payload;
    return(
      <div>
        <CommentReplyBanner getReplies={this.handleFetch}
                            commentId={this.props.commentId}
                            replyList={replies}
                            user={this.props.authState.username}
                            authorImage={this.props.authState.image}
                            IsAuth={this.props.authState.IsAuth}
                            authorUsername={this.props.authState.username}
                            handleChange={this.handleChange}
                            handleSubmit={this.handleSubmit}
                            value={this.state.reply_body}
                            deleteReply={this.handleDelete}
          />
          <ToastContainer/>
      </div>

    )
  }
}
export const mapStateToProps = state => {
  return{
    commentReplyState:state.commentReplies,
    authState:state.Auth
  }


}

export default withRouter(connect(mapStateToProps ,{ getCommentReplies, createCommentReplyAction, deleteCommentReplyAction }) (CommentReplyView));
