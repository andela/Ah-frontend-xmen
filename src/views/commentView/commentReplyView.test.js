import React from 'react';
import { shallow, mount} from 'enzyme';
import {CommentReplyView,mapStateToProps} from './commentReplyView';
import {
  createCommentReplyAction, deleteCommentReplyAction,
  getCommentReplies, updateCommentReplyAction
} from '../../actions/commentActions/commentReplyActions';


const props = {
  commentId:'1',
  authState: {
    username: "",
    image:"",
    IsAuth:true,
  },
  commentReplyState:{
    payload:[],
    isSuccessful:false,
    replyDeleteSuccess:false
  },
  match:{
    params:{
      slug:""
    }
  },
  getCommentReplies: jest.fn(),
  createCommentReplyAction: jest.fn(),
  deleteCommentReplyAction: jest.fn(),
  updateCommentReplyAction: jest.fn()
}
jest.useFakeTimers();
describe('<CommentReplyView',() => {

  it('should render without crashing', () => {
    let wrapper = shallow(<CommentReplyView {...props} />)
    expect(wrapper).toMatchSnapshot();
  });

  it('should reset reply body state on successful creation of reply', () => {
    let wrapper = mount(<CommentReplyView {...props} />)
    wrapper.setState({
      reply_body:"A value"
    })

    wrapper.setProps({
      commentState:{
        isSuccessful: true
      }
    })
    wrapper.update()


  });
  it('should call updateCommentReplyAction ', () => {
    let wrapper = shallow(<CommentReplyView {...props} />);
    const e ={
      preventDefault:()=>{},
      target:{
        id:1
      }
    }
    wrapper.instance().handleUpdateSubmit(e)
    expect(props.updateCommentReplyAction).toBeCalled()
  })

  it('should  handleUpdateChange', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
    const e ={
      target: {
        value:"update body"
      }
    }
    wrapper.instance().handleUpdateChange(e)
    expect(wrapper.instance().state.updateBody).toEqual("update body")

  });
  it('should call deleteCommentReplyAction', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
    const e ={
      target:{
        id:1
      }
    }
    wrapper.instance().handleDelete(e);
    expect(props.deleteCommentReplyAction).toBeCalled();

  });
  it('should call getCommentReplies', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
     wrapper.instance().handleFetch()
    expect(props.getCommentReplies).toBeCalled()

  })
  it('should call createCommentReplyAction', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
    const e ={
      preventDefault: () => {}
    }
    wrapper.instance().handleSubmit(e)
    jest.runAllTimers();
    expect(props.createCommentReplyAction).toBeCalled()
  });

  it('should handleChange', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
    const e = {
      target:{
        value:"change value"
      }
    }
    wrapper.instance().handleChange(e)
    expect(wrapper.instance().state.reply_body).toEqual("change value")
  });
  it('should open collapsible', () => {
    let wrapper = mount(<CommentReplyView {...props} />);
    const e = {
      target:{
        id: 1
      }
    }
    wrapper.instance().handleOpen(e)
    // expect(wrapper.instance().state.reply_body).toEqual("change value")
  });

  it('should map state to props', () => {
    expect(mapStateToProps({})).toEqual({"authState": undefined, "commentReplyState": undefined})
  });

})
