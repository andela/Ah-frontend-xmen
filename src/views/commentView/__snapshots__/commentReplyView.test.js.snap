// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentReplyView should render without crashing 1`] = `ShallowWrapper {}`;
